added few info on non-page components (#15198)
gatsbyjs/gatsby: docs/docs/building-with-components.md

sites. To see full working examples, check out the
[examples directory](https://github.com/gatsbyjs/gatsby/tree/master/examples) in
the Gatsby repo.

> ### Querying data in components using StaticQuery

Reviewer: I think this should be more about what "non-page components" are than how to use StaticQuery. The reason to write about it is to support StaticQuery docs, but this one should be focused more on the type of component that StaticQuery can be used in.

Gatsby v2 introduces StaticQuery, a new API that allows components to retrieve data via GraphQL query.

> <br>

Reviewer: ```suggestion

```
These break tags aren't necessary in Markdown.

sites. To see full working examples, check out the
[examples directory](https://github.com/gatsbyjs/gatsby/tree/master/examples) in
the Gatsby repo.

### Non-page components

> A Non-page component is embedded inside some other component forming a hierarchy that can get quite deep

Reviewer: ```suggestion
A Non-page component is one that's embedded inside some other component, forming a component hierarchy. An example would be a Header component that's included in multiple page components.
```

> Gatsby uses GraphQL to enable components to declare the data they need.Using StaticQuery, you can colocate a component with its data. 

Reviewer: ```suggestion
Gatsby uses GraphQL to enable components to declare the data they need. Using the [StaticQuery](/docs/static-query/) component or [useStaticQuery hook](/docs/use-static-query/), you can colocate a non-page component with its data.
```
